syntax = "proto3";
package message.v1;


service MessageService {
	// Append a message; returns assigned offset
	rpc Append(AppendReq) returns (AppendResp);
	// List history from an offset
	rpc List(ListReq) returns (stream Msg);
	// Subscribe to new messages for a room (server-stream)
	rpc Subscribe(RoomId) returns (stream Msg);
	// Query next offset (for resume helpers)
	rpc NextOffset(RoomId) returns (Offset);
}


message RoomId { string room_id = 1; }
message Offset { int64 value = 1; }


message AppendReq { string room_id = 1; string user_id = 2; string text = 3; string idempotency_key = 4; }
message AppendResp { bool success = 1; int64 offset = 2; string error = 3; }


message ListReq { string room_id = 1; int64 from_offset = 2; int32 limit = 3; }
message Msg { string room_id = 1; string user_id = 2; string text = 3; int64 offset = 4; int64 ts_ms = 5; }